### PR TITLE
GEOPY-1551: Use scikit-image.measure instead of matplotlib contours 

### PR DESCRIPTION
**Pull request by benk-mira**
**GEOPY-1551 - Use scikit-image.measure instead of matplotlib contours **

### CODE REVIEW

File: curve_apps/contours/driver.py [hunk 1]
```diff
@@ -49,60 +51,130 @@ def make_curve(self):
...
-        return contours
+
+        return (
+            np.vstack(vertices),
+            np.vstack(edges).astype("uint32"),
+            np.hstack(values),
+        )
+
+    @staticmethod
+    def image_to_grid(image, grid) -> Callable:
```
Comment (domfournier) [inline curve_apps/contours/driver.py:132]: Can we rename this to something more descriptive, like "interp_grid_xy"? Can we also turn it into a function instead - will be easier to migrate out later.

Comment (domfournier) [inline curve_apps/contours/driver.py:132]: Type inputs

File: curve_apps/utils.py [hunk 2]
```diff
@@ -23,6 +24,43 @@
 from curve_apps.trend_lines.params import TrendLineDetectionParameters
 
 
+def interp_to_grid(
+    entity: ObjectBase, values: np.ndarray, resolution, max_distance
```
Comment (domfournier) [inline curve_apps/utils.py:28]:
```suggestion
    entity: ObjectBase, values: np.ndarray, resolution: float, max_distance: float
```

File: curve_apps/utils.py [hunk 3]
```diff
@@ -23,6 +24,43 @@
...
+    Interpolate values into a regular grid based on entity locations.
+
+    :param entity: Geoh5py object with locations data.
+    :param values: Data to be interpolated to grid.
+    :param resolution: Grid resolution
+    :param max_distance: Maximum distance used in weighted average.
+    """
+
+    grid = []
+    for i in np.arange(2):
```
Comment (domfournier) [inline curve_apps/utils.py:40]: Surprized that linters didn't complain about single letter variable.

File: curve_apps/contours/driver.py [hunk 4]
```diff
@@ -49,60 +51,130 @@ def make_curve(self):
...
+        for contour in contour_list:
+
+            coords = measure.find_contours(data, contour, fully_connected="low")
+            if not coords:
+                continue
+
+            for coord in coords:
+                segment_vertices = interp(coord[:, 1], coord[:, 0])
+                nv = len(segment_vertices)
+                segment_edges = [[k, k + 1] for k in range(v0, v0 + (nv - 1))]
```
Comment (domfournier) [inline curve_apps/contours/driver.py:113]:
```suggestion
                segment_edges = np.c_[np.arange(nv-1),  np.arange(1, nv)] + v0
```

More efficient than looping
image

**Review by domfournier — COMMENTED** [review]
Looks great. See comments.

File: curve_apps/contours/driver.py [hunk 5]
```diff
@@ -49,60 +51,130 @@ def make_curve(self):
...
+        vertices, edges, values = [], [], []
+        v0 = 0
+        for contour in contour_list:
+
+            coords = measure.find_contours(data, contour, fully_connected="low")
+            if not coords:
+                continue
+
+            for coord in coords:
+                segment_vertices = interp(coord[:, 1], coord[:, 0])
```
Comment (domfournier) [inline curve_apps/contours/driver.py:111]: Also, we should probably stash the interpolation callable instead of re-generating everytime in the loop.

**Review by domfournier — COMMENTED** [review]
One more

File: curve_apps/contours/driver.py [hunk 6]
```diff
@@ -49,60 +51,130 @@ def make_curve(self):
...
+        row = np.arange(image.shape[0])
+        x_interp = interp1d(col, np.unique(grid[0]))
+        y_interp = interp1d(row, np.unique(grid[1]))
+
+        def interpolator(col, row):
+            return np.c_[x_interp(col), y_interp(row)]
+
+        return interpolator
+
+    # @staticmethod
```
[the remark this answers is not in the document]
Comment (benk-mira) [reply curve_apps/contours/driver.py:148]: ah, shoot.  I forgot to add the grid2d handling..   I'll cleanup once I've handled all the cases

[the remark this answers is not in the document]
Comment (benk-mira) [reply curve_apps/contours/driver.py:168]: done

File: curve_apps/contours/driver.py (hunk 1 above)
Comment (benk-mira) [reply curve_apps/contours/driver.py]: image_to_grid_coordinate_transfer is now a util

File: curve_apps/utils.py (hunk 3 above)
Comment (benk-mira) [reply curve_apps/utils.py]: na na na boo boo linter.  I think it's totally reasonable to have single letter iterables

**Comment by benk-mira** [comment]
Finally working for both grid and other generic types.  We might want to clip the curves to the source objects.. Kind of ran out of time this sprint.  I'm fine to fix here or new issue. up to you.

File: curve_apps/utils.py [hunk 7]
```diff
@@ -12,17 +12,75 @@
 from __future__ import annotations
 
 import re
+from collections.abc import Callable
 
 import numpy as np
+from geoapps_utils.numerical import weighted_average
 from geoh5py.objects import Curve, Grid2D, ObjectBase, Points, Surface
 from matplotlib.contour import ContourSet
```
Comment (domfournier) [inline curve_apps/utils.py:20]: Can remove this, and also remove matplotlib from the toml

File: curve_apps/utils.py [hunk 8]
```diff
@@ -12,17 +12,75 @@
...
+        threshold=resolution / 2.0,
+        n=8,
+        max_distance=max_distance,
+    )
+    values = values[0].reshape(x.shape)
+
+    return grid, values
+
+
 def extract_data(contours: ContourSet) -> tuple[list, list, list]:
```
Comment (domfournier) [inline curve_apps/utils.py:84]: This function can go.

**Review by domfournier — COMMENTED** [review]
looks good and tested well. Just a bit more clean ups

File: curve_apps/utils.py (hunk 8 above)
Comment (benk-mira) [reply curve_apps/utils.py]: done

File: curve_apps/utils.py (hunk 7 above)
Comment (benk-mira) [reply curve_apps/utils.py]: done

**Review by domfournier — APPROVED** [review]

**Pull request merged**